File: problem3/p3.c
/*
 * p3.c
 *
 * Zeke Reyna
 *
 * A simple function that finds the largest prime factor of a number.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define NUM 600851475143

int main()
{
	int largest_prime;
	unsigned long long num = NUM;
	long bound = (long) sqrt( num );
	int prime_factors[ bound ];
	int i, j;

	largest_prime = 1;
	prime_factors[0] = 1;
	prime_factors[1] = -1; //-1 used as null terminator

	for( i = 2; i < bound; i++ ) {
		if( num % i == 0 ) {
			for( j = 1; i % prime_factors[j] != 0; j++ )
				;

			if( prime_factors[j] == -1 ) {
				prime_factors[j] = i;
				prime_factors[++j] = -1;
			}
		}
	}

	for( i = 0; prime_factors[i] != -1; i++ )
		;

	largest_prime = prime_factors[--i];

	printf( "\nThe largest prime factor" );
	printf( " of %llu is %d\n\n", num, largest_prime );
	return 0;
}

File: problem3/vers_java/P3.java
/*
 * P3.java
 *
 * Zeke Reyna
 *
 * A simple function that finds the largest
 * prime factor of a number
 */

import java.util.*;
import java.math.*;

public class P3
{

        public static void print(Object line)
        {
                System.out.print(line);
        }

        public static BigInteger sqrt(BigInteger num)
        {
                BigInteger div = BigInteger.ZERO.setBit(num.bitLength() / 2);
                BigInteger div2 = div;

                for (;;) {
                        BigInteger temp = div.add(num.divide(div)).shiftRight(1);
                        if (temp.equals(div) || temp.equals(div2))
                                return num;
                        div2 = div;
                        div = temp;
                }
        }

        public static void main( String args[] )
        {
                BigInteger i;
                BigInteger largest_prime;
                BigInteger num;
                BigInteger bound;
                BigInteger temp;
                boolean isPrime;

                ArrayList<BigInteger> prime_factors = new ArrayList<BigInteger>();
                prime_factors.add(new BigInteger("2"));

                if (args.length == 1)
                        num = new BigInteger(args[0]);
                else
                        num = new BigInteger("100");

                //bound = num.divide(new BigInteger("2"));
                bound = sqrt(num);

                for (i = new BigInteger("2");
                    !i.equals(num.add(BigInteger.ONE));
                            i = i.add(BigInteger.ONE)) {

                        if (num.mod(i).equals(BigInteger.ZERO)) {

                                isPrime = true;
                                for (BigInteger j : prime_factors) {

                                        if (i.mod(j).equals(BigInteger.ZERO))
                                                isPrime = false;
                                }

                                if (isPrime)
                                        prime_factors.add(i);
                        }
                }

                largest_prime = prime_factors.get(prime_factors.size() - 1);

                print( "\nThe largest prime factor of " + num );
                print( " is " + largest_prime + "\n\n" );
        }
}
